# -*- coding: utf-8 -*-
# Generated by Django 1.11.6 on 2018-05-16 15:22
from __future__ import unicode_literals

import array_field_select.fields
from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('panels', '0052_auto_20180515_1239'),
    ]

    operations = [
        migrations.AlterField(
            model_name='region',
            name='type_of_effects',
            field=array_field_select.fields.ArrayField(base_field=models.CharField(choices=[('transcript_ablation', 'High: A feature ablation whereby the deleted region includes a transcript feature'), ('splice_acceptor_variant', "High: A splice variant that changes the 2 base region at the 3' end of an intron"), ('splice_donor_variant', "High: A splice variant that changes the 2 base region at the 5' end of an intron"), ('stop_gained', 'High: A sequence variant whereby at least one base of a codon is changed, resulting in a premature stop codon, leading to a shortened transcript'), ('frameshift_variant', 'High: A sequence variant which causes a disruption of the translational reading frame, because the number of nucleotides inserted or deleted is not a multiple of three'), ('stop_lost', 'High: A sequence variant where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript'), ('start_lost', 'High: A codon variant that changes at least one base of the canonical start codon'), ('transcript_amplification', 'High: A feature amplification of a region containing a transcript'), ('inframe_insertion', 'Moderate: An inframe non synonymous variant that inserts bases into in the coding sequence'), ('inframe_deletion', 'Moderate: An inframe non synonymous variant that deletes bases from the coding sequence'), ('missense_variant', 'Moderate: A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved'), ('protein_altering_variant', 'Moderate: A sequence_variant which is predicted to change the protein encoded in the coding sequence'), ('splice_region_variant', 'Low: A sequence variant in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8 bases of the intron'), ('incomplete_terminal_codon_variant', 'Low: A sequence variant where at least one base of the final codon of an incompletely annotated transcript is changed'), ('start_retained_variant', 'Low: A sequence variant where at least one base in the start codon is changed, but the start remains'), ('stop_retained_variant', 'Low: A sequence variant where at least one base in the terminator codon is changed, but the terminator remains'), ('synonymous_variant', 'Low: A sequence variant where there is no resulting change to the encoded amino acid'), ('coding_sequence_variant', 'Modifier: A sequence variant that changes the coding sequence'), ('mature_miRNA_variant', 'Modifier: A transcript variant located with the sequence of the mature miRNA'), ('5_prime_UTR_variant', "Modifier: A UTR variant of the 5' UTR"), ('3_prime_UTR_variant', "Modifier: A UTR variant of the 3' UTR"), ('non_coding_transcript_exon_variant', 'Modifier: A sequence variant that changes non-coding exon sequence in a non-coding transcript'), ('intron_variant', 'Modifier: A transcript variant occurring within an intron'), ('NMD_transcript_variant', 'Modifier: A variant in a transcript that is the target of NMD'), ('non_coding_transcript_variant', 'Modifier: A transcript variant of a non coding RNA gene'), ('upstream_gene_variant', "Modifier: A sequence variant located 5' of a gene"), ('downstream_gene_variant', "Modifier: A sequence variant located 3' of a gene"), ('TFBS_ablation', 'Modifier: A feature ablation whereby the deleted region includes a transcription factor binding site'), ('TFBS_amplification', 'Modifier: A feature amplification of a region containing a transcription factor binding site'), ('TF_binding_site_variant', 'Modifier: A sequence variant located within a transcription factor binding site'), ('regulatory_region_ablation', 'Moderate: A feature ablation whereby the deleted region includes a regulatory region'), ('regulatory_region_amplification', 'Modifier: A feature amplification of a region containing a regulatory region'), ('feature_elongation', 'Modifier: A sequence variant that causes the extension of a genomic feature, with regard to the reference sequence'), ('regulatory_region_variant', 'Modifier: A sequence variant located within a regulatory region'), ('feature_truncation', 'Modifier: A sequence variant that causes the reduction of a genomic feature, with regard to the reference sequence'), ('intergenic_variant', 'Modifier: A sequence variant located in the intergenic region, between genes')], max_length=128), help_text='Press CTRL or CMD button to select multiple effects', size=None, verbose_name='Consequence Types'),
        ),
        migrations.AlterField(
            model_name='region',
            name='type_of_variants',
            field=models.CharField(choices=[('Small', 'Small'), ('SV', 'SV'), ('CNV', 'CNV')], max_length=32, verbose_name='Variant type'),
        ),
        migrations.AlterField(
            model_name='trackrecord',
            name='issue_type',
            field=models.CharField(choices=[('Created', 'Created'), ('NewSource', 'Added New Source'), ('RemovedSource', 'Removed Source'), ('ChangedGeneName', 'Changed Gene Name'), ('SetPhenotypes', 'Set Phenotypes'), ('SetModelofInheritance', 'Set Model of Inheritance'), ('ClearSources', 'Clear Sources'), ('SetModeofPathogenicity', 'Set mode of pathogenicity'), ('GeneClassifiedbyGenomicsEnglandCurator', 'Gene classified by Genomics England curator'), ('EntityClassifiedbyGenomicsEnglandCurator', 'Entity classified by Genomics England curator'), ('SetModeofInheritance', 'Set mode of inheritance'), ('SetPenetrance', 'Set penetrance'), ('SetPublications', 'Set publications'), ('ApprovedGene', 'Approved Gene'), ('ApprovedEntity', 'Approved Entity'), ('GelStatusUpdate', 'Gel Status Update'), ('UploadGeneInformation', 'Upload gene information'), ('RemovedTag', 'Removed Tag'), ('AddedTag', 'Added Tag'), ('ChangedName', 'Changed Name'), ('ChangedSTRName', 'Changed STR Name'), ('ChangedChromosome', 'Changed Chromosome'), ('ChangedPosition37', 'Changed GRCh37'), ('ChangedPosition38', 'Changed GRCh38'), ('ChangedNormalRepeats', 'Changed Normal Number of Repeats'), ('ChangedPathogenicRepeats', 'Changed Pathogenic Number of Repeats'), ('RemovedGene', 'Removed Gene from the STR'), ('ChangedRepeatedSequence', 'Changed Repeated Sequence'), ('ChangedEffectTypes', 'Changed Effect Types'), ('ChangedVariantType', 'Changed Variant Types')], max_length=512),
        ),
    ]
