# Generated by Django 2.1.3 on 2018-11-30 14:20

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('panels', '0067_auto_20180904_1246'),
    ]

    operations = [
        migrations.AlterModelOptions(
            name='evidence',
            options={'ordering': ['-created']},
        ),
        migrations.AlterField(
            model_name='evaluation',
            name='current_diagnostic',
            field=models.BooleanField(blank=True, default=False),
        ),
        migrations.AlterField(
            model_name='region',
            name='haploinsufficiency_score',
            field=models.CharField(blank=True, choices=[('', ''), ('3', 'Sufficient evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('2', 'Emerging evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('1', 'Little evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('0', 'No evidence to suggest that dosage sensitivity is associated with clinical phenotype'), ('40', 'Dosage sensitivity unlikely'), ('30', 'Gene associated with autosomal recessive phenotype')], max_length=2, null=True),
        ),
        migrations.AlterField(
            model_name='region',
            name='triplosensitivity_score',
            field=models.CharField(blank=True, choices=[('', ''), ('3', 'Sufficient evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('2', 'Emerging evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('1', 'Little evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('0', 'No evidence to suggest that dosage sensitivity is associated with clinical phenotype'), ('40', 'Dosage sensitivity unlikely'), ('30', 'Gene associated with autosomal recessive phenotype')], max_length=2, null=True),
        ),
        migrations.AlterField(
            model_name='trackrecord',
            name='issue_type',
            field=models.CharField(choices=[('Created', 'Created'), ('NewSource', 'Added New Source'), ('RemovedSource', 'Removed Source'), ('ChangedGeneName', 'Changed Gene Name'), ('SetPhenotypes', 'Set Phenotypes'), ('SetModelofInheritance', 'Set Mode of Inheritance'), ('ClearSources', 'Clear Sources'), ('SetModeofPathogenicity', 'Set mode of pathogenicity'), ('GeneClassifiedbyGenomicsEnglandCurator', 'Gene classified by Genomics England curator'), ('EntityClassifiedbyGenomicsEnglandCurator', 'Entity classified by Genomics England curator'), ('SetModeofInheritance', 'Set mode of inheritance'), ('SetPenetrance', 'Set penetrance'), ('SetPublications', 'Set publications'), ('ApprovedGene', 'Approved Gene'), ('ApprovedEntity', 'Approved Entity'), ('GelStatusUpdate', 'Status Update'), ('UploadGeneInformation', 'Upload gene information'), ('RemovedTag', 'Removed Tag'), ('AddedTag', 'Added Tag'), ('ChangedName', 'Changed Name'), ('ChangedSTRName', 'Changed STR Name'), ('ChangedChromosome', 'Changed Chromosome'), ('ChangedPosition37', 'Changed GRCh37'), ('ChangedPosition38', 'Changed GRCh38'), ('ChangedNormalRepeats', 'Changed Normal Number of Repeats'), ('ChangedPathogenicRepeats', 'Changed Pathogenic Number of Repeats'), ('RemovedGene', 'Removed Gene from the STR'), ('ChangedRepeatedSequence', 'Changed Repeated Sequence'), ('ChangedEffectTypes', 'Changed Effect Types'), ('ChangedVariantType', 'Changed Variant Types'), ('ChangedHaploinsufficiencyScore', 'Changed Haploinsufficiency Score'), ('ChangedTriplosensitivityScore', 'Changed Triplosensitivity Score'), ('ChangedRequiredOverlapPercentage', 'Changed Required Overlap Percentage')], max_length=1024),
        ),
    ]
